src/libfetchers/git-utils.cc

```diff
@@ -322,8 +322,11 @@ struct GitRepoImpl : GitRepo, std::enable_shared_from_this<GitRepoImpl>
 
             for (size_t n = 0; n < git_commit_parentcount(commit->get()); ++n) {
                 git_commit * parent;
-                if (git_commit_parent(&parent, commit->get(), n))
-                    throw Error("getting parent of Git commit '%s': %s", *git_commit_id(commit->get()), git_error_last()->message);
+                if (git_commit_parent(&parent, commit->get(), n)) {
+                    // If the parent commit cannot be found (e.g., in a worktree derived from a shallow clone),
+                    // return the number of commits found so far. This matches the behavior of `git rev-list --all`.
+                    return done.size();
```

Comment:
We can't continue with incorrect information, as that breaks evaluation purity.
The correct thing to do is to fetch the commit history. That can still be a shallow fetch, with `git fetch --filter=tree:none`, i.e. fetching about the minimum amount of objects that is needed for the count.

Alternatively / additionally the impact of this error could be reduced by computing the `revCount` attribute lazily. That would also increase performance in the general case where `revCount` isn't evaluated.

Reply:
I just checked the history and before we switched to libgit2 we also used `git rev-list --count <commit>`:

```
$ git rev-list --count HEAD
2
```

So this is a regression and I don't see why we should not stick to the semantics of git.

Reply:
Git CLI semantics are bad. We don't support impure things like `export-subst` either.
Fortunately this is still an experimental feature, so we have no reason to choose impure behavior.

> potentially expensive on git infrastructure (turning a shallow clone into a full one).

It does not do that. It only fetches the little commit objects, and no trees or blobs.

> nixpkgs ci

Does nixpkgs ci need a value for `revCount`? Otherwise laziness, as proposed, is another way out.

Reply:
Sorry I don't think what you are asking is possible to implement? Where should this fetch from? There doesn't have to be a remote set or there might be multiple remote to choose from. The remote might require authentication and network that is not available to the current user that runs nix (i.e. something someone might want to do for security reasons in CI systems).

Reply:
This took 50s on nixpkgs with a gigabit connection and downloaded.

```
% git fetch --filter=tree:0 --depth=99999 origin master
remote: Enumerating objects: 806036, done.
remote: Counting objects: 100% (806036/806036), done.
remote: Compressing objects: 100% (767969/767969), done.
remote: Total 806036 (delta 38067), reused 806031 (delta 38067), pack-reused 0 (from 0)
Receiving objects: 100% (806036/806036), 336.58 MiB | 43.36 MiB/s, done.
Resolving deltas: 100% (38067/38067), done.
From https://github.com/NixOS/nixpkgs
 * branch                  master     -> FETCH_HEAD
noglob hub fetch --filter=tree:0 --depth=99999 origin master  12.38s user 8.62s system 45% cpu 46.003 total
```

For comparison the original shallow clone only took 9s.

Reply:
Now it no longer changes anything about the revCount but throws a different error message.